src/main/java/mate.academy/lib/Injector.java

```diff
@@ -1,13 +1,77 @@
 package mate.academy.lib;
 
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Field;
+import java.util.HashMap;
+import java.util.Map;
+import mate.academy.service.FileReaderService;
+import mate.academy.service.ProductParser;
+import mate.academy.service.ProductService;
+import mate.academy.service.impl.FileReaderServiceImpl;
+import mate.academy.service.impl.ProductParserImpl;
+import mate.academy.service.impl.ProductServiceImpl;
+
 public class Injector {
-    private static final Injector injector = new Injector();
+    private static final Map<Class<?>, Class<?>> interfacesImplementation;
+    private static final Injector injector;
+    private Map<Class<?>, Object> instances = new HashMap<>();
+
+    static {
+        interfacesImplementation = Map.of(
+                FileReaderService.class, FileReaderServiceImpl.class,
+                ProductParser.class, ProductParserImpl.class,
+                ProductService.class, ProductServiceImpl.class
+        );
+        injector = new Injector();
+    }
 
     public static Injector getInjector() {
         return injector;
     }
 
     public Object getInstance(Class<?> interfaceClazz) {
-        return null;
+        Object newInstance = null;
+        Class<?> clazz = findImplementation(interfaceClazz);
+        if (!clazz.isAnnotationPresent(Component.class)) {
+            throw new RuntimeException("Class must be annotated with @Component");
+        }
+        Field[] declaredFields = clazz.getDeclaredFields();
+        for (Field field : declaredFields) {
+            if (field.isAnnotationPresent(Inject.class)) {
+                Object fieldInstance = getInstance(field.getType());
+                newInstance = createNewInstance(clazz);
+                try {
+                    field.setAccessible(true);
+                    field.set(newInstance, fieldInstance);
+                } catch (IllegalAccessException e) {
+                    throw new RuntimeException("Can't initialize field value. "
+                            + "Class: " + clazz.getName() + ". "
+                            + "Field: " + field.getName());
+                }
+            }
+        }
+        if (newInstance == null) {
+            newInstance = createNewInstance(clazz);
+        }
+        return newInstance;
+    }
+
+    private Object createNewInstance(Class<?> clazz) {
+        if (instances.containsKey(clazz)) {
+            return instances.get(clazz);
+        }
+        try {
+            Constructor<?> constructor = clazz.getConstructor();
+            Object instance = constructor.newInstance();
+            instances.put(clazz, instance);
+            return instance;
+        } catch (ReflectiveOperationException e) {
+            throw new RuntimeException("Can't create a new instance" + clazz.getName());
```

Comment:
When you throw an exception add an informative message to it. Also, don’t forget to add an exception you’re catching in the catch block to the RuntimeException object you throw

Bad example: 
} catch(Exception e) {
throw new SomeException("Can't initialize object with Injector");
}
Good example: 
} catch(Exception e) {
throw new SomeException("Injection failed, missing @Component annotaion on the class " + someInfoAboutClass, e);
}